src/decompression.jl

```diff
@@ -701,54 +701,72 @@ end
-    =#
-    (; A, col_color_ind, row_color_ind) = result
-    m, n = size(A)
-    R = Base.promote_eltype(Br, Bc)
-    if eltype(result.Br_and_Bc) == R
-        Br_and_Bc = result.Br_and_Bc
+"""
+    JoinCompressed{R<:Real}
+
+For a bicoloring of an original matrix `A` of size `(m, n)`, we build an augmented matrix `A_and_Aᵀ = [0 Aᵀ; A 0]`.
+Symmetric column coloring is then performed on `A_and_Aᵀ`, and the column-wise compression of `A_and_Aᵀ` produces a matrix `Br_and_Bc`.
+In the case of bicoloring, `Br_and_Bc` is computed as two partial compressions: the row-wise compression `Br` (corresponding to `Aᵀ`) and the column-wise compression `Bc` (corresponding to `A`).
+
+For the symmetric decompression, we lazily reconstruct `Br_and_Bc` from `Br` and `Bc`, knowing that the symmetric colors (those making up `Br_and_Bc`) may appear in either a row of `Br`, a column of `Bc`, or both.
+The columns of the top part of `Br_and_Bc` (rows between `1` and `n`) are taken from the rows of `Br`, interleaved with zero columns whenever the current color has not been used to color any row.
+The columns of the bottom part of `Br_and_Bc` (rows between `n+1` and `n+m`) are taken from the columns of `Bc`, interleaved with zero columns whenever the current color has not been used to color any column.
+
+We use the dictionaries `col_color_ind` and `row_color_ind` to map colors obtained during star or acyclic coloring to row colors in `Br` and column colors in `Bc`.
+
+# Fields
+
+- `m::Int` : number of rows in `A` and `Bc`
+- `n::Int` : number of columns in `A` and `Br`
+- `c::Int` : number of colors used for the symmetric coloring
+- `Br::AbstractMatrix{R}` : row-wise compressed matrix
+- `Bc::AbstractMatrix{R}` : column-wise compressed matrix
+- `row_color_ind::Dict{Int,Int}` : dictionary mapping symmetric colors to row indices in `Br`
+- `col_color_ind::Dict{Int,Int}` : dictionary mapping symmetric colors to column indices in `Bc`
+"""
+struct JoinCompressed{R<:Real,M1<:AbstractMatrix{R},M2<:AbstractMatrix{R}} <:
+       AbstractMatrix{R}
+    m::Int
+    n::Int
+    Br::M1
+    Bc::M2
+    row_color_ind::Dict{Int,Int}
+    col_color_ind::Dict{Int,Int}
+end
+
+function Base.getindex(B::JoinCompressed, i::Integer, j::Integer)
+    (; n, Br, Bc, row_color_ind, col_color_ind) = B
+    if i ≤ n
+        return Br[row_color_ind[j], i]
     else
-        Br_and_Bc = similar(result.Br_and_Bc, R)
+        return Bc[i - n, col_color_ind[j]]
     end
-    fill!(Br_and_Bc, zero(R))
-    for c in axes(Br_and_Bc, 2)
-        if haskey(row_color_ind, c)  # some rows were colored with symmetric color c
-            copyto!(view(Br_and_Bc, 1:n, c), view(Br, row_color_ind[c], :))
-        end
-        if haskey(col_color_ind, c)  # some columns were colored with symmetric c
-            copyto!(view(Br_and_Bc, (n + 1):(n + m), c), view(Bc, :, col_color_ind[c]))
-        end
-    end
```

Comment:
Actually the previous storage wasn't so bad at all. If every color is used both for rows and for columns, there's no waste whatsoever. I think the performance comparison is not clear cut

Reply:
The waste is just that we duplicate memory Guillaume.
Why should we allocate `Br_and_Bc` and we have already `Br` and `Bc`.

Reply:
It is a clear cut in terms of storage, for the performance I don't know with the current implementation of dictionary.